File: lib/src/internal/routes/password.dart
import '/src/sdk/routes.dart';
import '/src/types/others.dart';
import '/src/types/responses.dart';
import '../http/descope_client.dart';
import '../http/responses.dart';
import 'shared.dart';

class Password implements DescopePassword {
  final DescopeClient client;

  Password(this.client);

  @override
  Future<AuthenticationResponse> signUp({required String loginId, required String password, SignUpDetails? details}) async {
    return (await client.passwordSignUp(loginId, password, details)).toAuthenticationResponse();
  }

  @override
  Future<AuthenticationResponse> signIn({required String loginId, required String password}) async {
    return (await client.passwordSignIn(loginId, password)).toAuthenticationResponse();
  }

  @override
  Future<void> update({required String loginId, required String newPassword, required String refreshJwt}) {
    return client.passwordUpdate(loginId, newPassword, refreshJwt);
  }

  @override
  Future<AuthenticationResponse> replace({required String loginId, required String oldPassword, required String newPassword}) async {
    return (await client.passwordReplace(loginId, oldPassword, newPassword)).convert();
  }

  @override
  Future<void> sendReset({required String loginId, String? redirectUrl}) {
    return client.passwordSendReset(loginId, redirectUrl);
  }

  @override
  Future<PasswordPolicy> getPolicy() async {
    return (await client.passwordGetPolicy()).convert();
  }
}

extension on PasswordPolicyServerResponse {
  PasswordPolicy convert() {
    return PasswordPolicy(minLength, lowercase, uppercase, number, nonAlphanumeric);
  }
}

File: ios/descope/Sources/descope/descope-swift-sdk/internal/routes/Password.swift

final class Password: DescopePassword {
    let client: DescopeClient
    
    init(client: DescopeClient) {
        self.client = client
    }
    
    func signUp(loginId: String, password: String, details: SignUpDetails?) async throws(DescopeError) -> AuthenticationResponse {
        return try await client.passwordSignUp(loginId: loginId, password: password, details: details).convert()
    }
    
    func signIn(loginId: String, password: String) async throws(DescopeError) -> AuthenticationResponse {
        return try await client.passwordSignIn(loginId: loginId, password: password).convert()
    }
    
    func update(loginId: String, newPassword: String, refreshJwt: String) async throws(DescopeError) {
        try await client.passwordUpdate(loginId: loginId, newPassword: newPassword, refreshJwt: refreshJwt)
    }
    
    func replace(loginId: String, oldPassword: String, newPassword: String) async throws(DescopeError) -> AuthenticationResponse {
        try await client.passwordReplace(loginId: loginId, oldPassword: oldPassword, newPassword: newPassword).convert()
    }
    
    func sendReset(loginId: String, redirectURL: String?) async throws(DescopeError) {
        try await client.passwordSendReset(loginId: loginId, redirectURL: redirectURL)
    }
    
    func getPolicy() async throws(DescopeError) -> PasswordPolicyResponse {
        return try await client.passwordGetPolicy().convert()
    }
}

private extension DescopeClient.PasswordPolicyResponse {
    func convert() throws(DescopeError) -> PasswordPolicyResponse {
        return PasswordPolicyResponse(minLength: minLength, lowercase: lowercase, uppercase: uppercase, number: number, nonAlphanumeric: nonAlphanumeric)
    }
}
